libraries/chain/transaction_context.cpp

```diff
@@ -454,6 +459,30 @@ namespace eosio { namespace chain {
       }
    }
 
+   void transaction_context::validate_account_cpu_usage( int64_t billed_cpu_time_us, int64_t account_cpu_limit )const {
+      if( account_cpu_limit < billed_cpu_time_us && !control.skip_trx_checks() ) {
```

Comment:
I expected this to be something like the following:

```cpp
   if( (billed_cpu_time_us > 0) && !control.skip_trx_checks() ) {
      bool    cpu_limited_by_account = (account_cpu_limit <= objective_duration_limit.count());
      int64_t cpu_limit =  (cpu_limited_by_account ? account_cpu_limit : objective_duration_limit.count());
      if( !cpu_limited_by_account && (billing_time_exception_code == block_cpu_usage_exceeded::code_value) ) {
         EOS_ASSERT( billed_cpu_time_us <= cpu_limit, block_cpu_usage_exceeded, ... );
      } else {
         if( cpu_limit_due_to_greylist ) {
            EOS_ASSERT( billed_cpu_time_us <= cpu_limit, greylist_cpu_usage_exceeded, ... );
         } else {
            EOS_ASSERT( billed_cpu_time_us <= cpu_limit, tx_cpu_usage_exceeded, ... );
         }
      }
   }
```

The current code would not throw an exception if the `billed_cpu_time_us` was greater than the `objective_duration_limit.count()` under the condition when the account's CPU limit was not the limiting factor (e.g. the estimate CPU usage was greater than the objective max limit of a transaction or it couldn't fit in the remainder of the block).

Reply:
The `objective_duration_limit` is associated with `block_cpu_usage_exceeded::code_value` unless either the chain configuration of `max_transaction_cpu_usage` changed (which is possible) OR the specified `max_cpu_usage_ms` on the transaction changed (which is not possible). I note this here because I was simplifying your conditions above and realized we still need to check `cpu_limit` and not `account_cpu_limit` in the `trx_cpu_usage_exceeded` case. Also we are not folding in `cpu_limit_due_to_greylist` correctly in the existing `validate_cpu_usage_to_bill()`.

Reply:
Reworked logic and consolidated with existing `validate_cpu_usage_to_bill`.